backend-comparison/src/persistence/base.rs

```diff
@@ -1,100 +1,105 @@
+    let file_path = cache_dir.join(file_name);
 
-        Self { results }
-    }
+    let records: Vec<BenchmarkRecord> = benches
+        .into_iter()
+        .map(|bench| BenchmarkRecord {
+            backend: B::name().to_string(),
+            device: format!("{:?}", device),
+            results: bench,
+        })
+        .collect();
 
-    /// Save the cache on disk.
-    fn save(&self, path: &PathBuf) {
-        if let Some(parent) = path.parent() {
-            create_dir_all(parent).expect("Unable to create directory");
-        }
-        let file = File::create(path).expect("Unable to create backend comparison file");
+    let file = fs::File::create(file_path).expect("Unable to create backend comparison file.");
+    serde_json::to_writer_pretty(file, &records).expect("Unable to save benchmark results.");
 
-        serde_json::to_writer_pretty(file, &self.results)
-            .expect("Unable to write to backend comparison file");
-    }
-
-    /// Update the cache with the given [benchmark results](BenchmarkResult).
-    ///
-    /// Assumes only that benches share the same backend and device.
-    /// It could run faster if we assumed they have the same git hash
-    fn update<B: Backend>(&mut self, device: &B::Device, benches: Vec<BenchmarkResult>) {
-        let backend_key = format!("{}-{:?}", B::name(), device);
-
-        for bench in benches {
-            let mut benchmark_op_results = self.results.remove(&bench.git_hash).unwrap_or_default();
-            let mut benchmark_backend_results =
-                benchmark_op_results.remove(&bench.name).unwrap_or_default();
-
-            let mut stamped_benchmarks = benchmark_backend_results
-                .remove(&backend_key)
-                .unwrap_or_default();
+    Ok(records)
+}
 
-            stamped_benchmarks.insert(bench.timestamp, bench.durations.durations);
-            benchmark_backend_results.insert(backend_key.clone(), stamped_benchmarks);
-            benchmark_op_results.insert(bench.name, benchmark_backend_results);
-            self.results.insert(bench.git_hash, benchmark_op_results);
-        }
+impl Serialize for BenchmarkRecord {
+    /// Flatten all the fields when serializing, i.e. we remove the nesting
+    /// under "results" and "computed".
+    /// Also format the fields to be compliant with MongoDB naming conventions.
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("BenchmarkRecord", 11)?;
+        state.serialize_field("backend", &self.backend)?;
+        state.serialize_field("device", &self.device)?;
+        // Serialize fields of BenchmarkResult
+        state.serialize_field("rawDurations", &self.results.raw.durations)?;
+        state.serialize_field("numSamples", &self.results.raw.durations.len())?;
+        state.serialize_field("mean", &self.results.computed.mean.as_secs_f64())?;
```

Comment:
Serialize in the smallest time unit possible, since we can benchmark things that take microseconds.